lnd_test.go

```diff
@@ -5351,14 +5351,151 @@ func testInvoiceSubscriptions(net *lntest.NetworkHarness, t *harnessTest) {
+	go func() {
+		defer cancelFunc()
+		for {
+			select {
+			case <-quit:
+				return
+			default:
+				chanUpdate, err := chanUpdateClient.Recv()
+				select {
+				case <-quit:
+					return
+				default:
+				}
+
+				if err == io.EOF {
+					return
+				} else if err != nil {
+					select {
+					case errChan <- err:
+					case <-quit:
+					}
+					return
+				}
 
+				select {
+				case chanUpdates <- chanUpdate:
+				case <-quit:
+					return
+				}
+			}
+		}
+	}()
+
+	return channelSubscription{
+		updateChan: chanUpdates,
+		errChan:    errChan,
+		quit:       quit,
+	}
+}
+
+// verifyCloseUpdate is used to verify that a closed channel update is of the
+// expected type.
+func verifyCloseUpdate(chanUpdate *lnrpc.ChannelUpdate, inactiveNtfn bool,
+	force bool, forceType lnrpc.ChannelCloseSummary_ClosureType) error {
+
+	// We should receive one inactive and one closed notification
+	// for each channel.
+	switch update := chanUpdate.Channel.(type) {
+	case *lnrpc.ChannelUpdate_OpenChannel:
+		if !inactiveNtfn {
+			return fmt.Errorf("received inactive ntfn but " +
+				"should've been for closed channel")
+		}
+		if update.OpenChannel.Active {
+			return fmt.Errorf("channel active but shouldn't be")
+		}
+		if chanUpdate.Type != lnrpc.ChannelUpdate_INACTIVE_CHANNEL {
+			return fmt.Errorf("update type mismatch: "+
+				"expected %v, got %v",
+				lnrpc.ChannelUpdate_INACTIVE_CHANNEL,
+				chanUpdate.Type)
+		}
+	case *lnrpc.ChannelUpdate_ClosedChannel:
+		if inactiveNtfn {
+			return fmt.Errorf("received ntfn for closed channel " +
+				"but should've been for inactive channel")
+		}
+
+		if chanUpdate.Type !=
+			lnrpc.ChannelUpdate_CLOSED_CHANNEL {
+			return fmt.Errorf("update type mismatch: "+
+				"expected %v, got %v",
+				lnrpc.ChannelUpdate_CLOSED_CHANNEL,
+				chanUpdate.Type)
+		}
+
+		switch force {
+		case true:
+			if update.ClosedChannel.CloseType != forceType {
+				return fmt.Errorf("channel closure type mismatch: "+
+					"expected %v, got %v",
+					forceType,
+					update.ClosedChannel.CloseType)
+			}
+		case false:
+			if update.ClosedChannel.CloseType !=
+				lnrpc.ChannelCloseSummary_COOPERATIVE_CLOSE {
+				return fmt.Errorf("channel closure type "+
+					"mismatch: expected %v, got %v",
+					lnrpc.ChannelCloseSummary_COOPERATIVE_CLOSE,
+					update.ClosedChannel.CloseType)
+			}
+		}
+	default:
+		return fmt.Errorf("channel update channel of wrong type, "+
+			"expected inactive or closed channel, got %T",
+			update)
+	}
+
+	return nil
+}
+
+// testBasicChannelCreationAndUpdates tests multiple channel opening and closing,
+// and ensures that if a node is subscribed to channel updates they will be
+// received correctly for both cooperative and force closed channels.
+func testBasicChannelCreationAndUpdates(net *lntest.NetworkHarness, t *harnessTest) {
 	const (
 		numChannels = 2
+		timeout     = time.Duration(time.Second * 5)
 		amount      = maxBtcFundingAmount
 	)
+	ctxb := context.Background()
```

Comment:
style nit: define ctxb first line in test.

Reply:
fixed